### ISSUE TITLE
Reduce the import time of pytorch_lightning

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#12786]
## Proposed refactor

The current import time for the pytorch_lightning package on my machine is several seconds. There are some opportunities to improve this.

### Motivation

High import times have an impact on the development and debugging speed. 

### Benchmark

I benchmarked the import time in two environments:
1. Fresh environment with pytorch lightning installed, no extras.
2. My currrent environment, with many extras installed such as loggers, horovod, etc.

To measure the import time, I created a simple file which only imports pytorch_lightning:

```py
import pytorch_lightning as pl
```

Then I use the `importtime` package to measure the time and create a profile:

```
python -X importtime simple.py 2> import.log
```

Finally, I used tuna to visualize the profile:

```
tuna import.log
```

For the fresh environment, the total import time is <2 secs with the following profile:

<img width="1466" alt="image" src="https://user-images.githubusercontent.com/5495193/163724305-4f927dbd-cd52-4273-a37b-89782fd28104.png">

```
>  pip freeze  | grep torch                                            (clean-pl-env) 

pytorch-lightning==1.6.1
torch==1.11.0
torchmetrics==0.7.3
```

For a full development environment, the total import time is >4 seconds:

<img width="1466" alt="image" src="https://user-images.githubusercontent.com/5495193/163723996-8f600c91-0d0a-467c-a2d5-e54a73bbac36.png">

The times vary a bit between multiple runs. However, I have observed that the time is consistently higher when running in an environment with extras installed. Looking at the profiles, it looks like the origin of a large waste of time is coming from our 

`pytorch_lightning.utilities.imports` module where we evaluate some constants at import time:

https://github.com/PyTorchLightning/pytorch-lightning/blob/ae3226ced96e2bc7e62f298d532aaf2290e6ef34/pytorch_lightning/utilities/imports.py#L98-L124

It looks like if a 3rd party package is installed and takes a long time to import, this time gets added to our loading time as well, even if the package never ends up being used. This is because our `_module_available` and `_package_available` implementations attempt to [import the modules](https://github.com/PyTorchLightning/pytorch-lightning/blob/ae3226ced96e2bc7e62f298d532aaf2290e6ef34/pytorch_lightning/utilities/imports.py#L56) to check their availability. This can be very costly. 

### Pitch

Evaluate the import checks lazily. 
Convert 

```py
_X_AVAILALBE = _module_available("x")
```

to

```py
@lru_cache()
def _is_x_available() -> bool:
    return _module_available("x")
```

And investigate other opportunities to improve loading time given the above profile.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @borda @akihironitta

**Comment by awaelchli** [comment]
@carmocca @Borda Let me know if you think this issue is worth to consider. I observed this already in "Project X" with bigger impact due to its dependencies, and was able to fix it with the above approach.

**Comment by carmocca** [comment]
Thanks for the thoughtful analysis! We should definitely do this

**Comment by Atharva-Phatak** [comment]
@carmocca Can I work on this ? I am more than happy to contribute.

**Comment by carmocca** [comment]
Sure @Atharva-Phatak, you can do one PR per import.

We can also do https://github.com/PyTorchLightning/pytorch-lightning/issues/11826 jointly

**Comment by Atharva-Phatak** [comment]
@carmocca quick questions in #11826 you have mentioned some libraries which needs to be moved. So in my PR should I only use the mentioned libraries right ?

Also can't I do all the libraries in one PR ? or would it be better if I do one PR per import. 

Please let me know

**Comment by awaelchli** [comment]
Jointly doing the lazy functions and the move to the files makes sense. 
Doing one PR per import is maybe a bit overkill. I would group them by locality. 
For example: the two deepspeed in one pr, all logger-related imports in one pr, all fairscale imports in one pr and so on. 
Thanks for the help.

**Comment by Atharva-Phatak** [comment]
@awaelchli This makes sense. I will start working on this :)

**Comment by Atharva-Phatak** [comment]
```
@lru_cache()
def _is_x_available() -> bool:
    return _module_available("x")
```
@awaelchli can we redefine the above function as, I was confused since above function looks like we need to implement it for every module "x". 
```
@lru_cache()
def _is_x_available(x : str) -> bool:
      """x: name of the library"""
     return _module_available(x)

Please let me know.

**Comment by awaelchli** [comment]
I don't have a strong preference here. @carmocca @Borda ?

The name as part of the function name makes it less error prone to use and probably easier to refactor the current imports, as it it can be renamed using a simple regex pattern everywhere.

**Comment by Borda** [comment]
why not directly?
```py
@lru_cache()
def _module_available(x : str) -> bool:
      # this is the original function

**Comment by Atharva-Phatak** [comment]
@Borda can you clarify a bit ? :) Sorry for such a beginner question

**Comment by awaelchli** [comment]
@Atharva-Phatak to clarify, what @Borda is saying is that the `@lru_cache()` decorator can be added to the main function that does the import check and doesn't have to be added to each individual one as I showed in my post above. 

This detail and the naming aside, the important point of this issue is that we want to evaluate module availability _lazily_.

**Comment by awaelchli** [comment]
Yes the `lru_cache` can be added there. But this is not the topic of this issue. The real work is in converting the `_X_AVAILABLE` constants to functions and evaluate them lazily, i.e., at the individual module level only when they are needed! 

Would it help if I made an example PR for a single import, and then you could work on the remaining ones? Appreciate the help.

**Comment by Atharva-Phatak** [comment]
@awaelchli Yes that will be very helpful if I can have a sample PR. Then I will do it for all the modules :)

Thanks for the help and thanks for being patient with me :)

**Comment by Atharva-Phatak** [comment]
Hi Thank you I will take a look :)

**Comment by plutasnyy** [comment]
@Atharva-Phatak How's it going? if you want I can help you with that ;)

**Comment by Atharva-Phatak** [comment]
@plutasnyy It's going good. I am just a little busy with my final exams :)

**Comment by Atharva-Phatak** [comment]
@carmocca I would be happy to, I made some mistakes in my last PR and you had to do extra work 😞 . I was a bit hesitant. 
Please let me know.

**Comment by carmocca** [comment]
Making mistakes is totally fine! We are thorough on reviews to make sure they are fixed before merge, but give it a shot and try doing the changes for another module :heart:

**Comment by awaelchli** [comment]
We started doing this for deepspeed in a different way than proposed here, but actually we haven't really solved the problem this way. 

We still have `_X_AVAILALBE = _RequirementAvailable("x")` which gets evaluated immediately on module import! The proposal here is to evaluate completely lazily everywhere. So I'm suggesting not to move forward with the started approach as this would only shift the problem from one place to another. As described in the proposal, we need to convert all

```py
_X_AVAILALBE = _module_available("x")
```

to something CALLABLE like this:

```py
def _is_x_available() -> bool:
    return _module_available("x")
```

so it can be called where needed only. I can make a PR for one of the loggers to illustrate this.

**Comment by Atharva-Phatak** [comment]
@awaelchli Thank you for the sample PR. It seems you moved the wandb import checks to the wandb file itself ? Is that how the refactoring is needed ? 

Please let me know :)

**Comment by justusschock** [comment]
@awaelchli I use the following pattern for those things: https://gist.github.com/justusschock/4146685578c995d9ac89c58eb10ae898 

Works pretty well for me. Feel free to reuse it :)

**Comment by carmocca** [comment]
@justusschock that's basically the same as `_RequirementAvailable`

**Comment by justusschock** [comment]
@carmocca it‘s definitely similar. However with _RequirementAvailable you can only check whether packages are available and versions match, but you cannot call arbitrary functions or anything from those packages. I originally thought about that for lazily checking whether mps devices are available as you van virtually execute any code

**Comment by carmocca** [comment]
It would be interesting to repeat this analysis now that `lightning` imports all app, fabric, and pl packages

**Comment by Borda** [comment]
@awaelchli, do you have a script for it to share? :rabbit:

**Comment by awaelchli** [comment]
@Borda The details of how to run the benchmark is in the issue description. Here are the latest results. 

**Minimal**
Fresh environment, 
`pip install lightning`, nothing else:

![import_minimal](https://user-images.githubusercontent.com/5495193/226070847-c4d2c9a1-4933-4e43-a40f-97b2d7fc0ac9.png)


**Full**
Fresh environment
`pip install lightning deepspeed wandb neptune mlflow comet-ml tensorboard jsonargparse[signatures]`

![import_full](https://user-images.githubusercontent.com/5495193/226070922-18146039-6781-4df7-9397-d2e94c9b48f0.png)



These are just screenshots. If it is hard to see, here are the raw files:

[import_minimal.log](https://github.com/Lightning-AI/lightning/files/11007002/import_minimal.log)
[import_full.log](https://github.com/Lightning-AI/lightning/files/11007003/import_full.log)

You can download them, and inspect like so by zooming in in the browser:

```
pip install tuna
tuna import_full.log
```

In summary, for the bare version we can see we are at a roughly 1sec import time. The dominating part is the lightning_app side. For the full version with all optional dependencies installed, we are at ~2sec import time and we see that the wandb and mlflow are standing out. There might be an opportunity to optimize there. Overall, Lightning 2.0 has already improved the situation greatly, so nice progress.

**Comment by carmocca** [comment]
Thank you! This is super valuable to know which items we should tackle first

**Comment by davidgilbertson** [comment]
I came here to log an issue with almost the exact same screenshot from tuna. It's excellent to see a bunch of smart people are already on the case. Good stuff!

**Comment by awaelchli** [comment]
I reran the benchmark now that we have dropped the top-level import of the app module in https://github.com/Lightning-AI/lightning/pull/18386. 

**Minimal**
Fresh environment,
`pip install lightning`, nothing else:
![minimal-install](https://github.com/Lightning-AI/lightning/assets/5495193/17cc39fa-acfd-47c2-a9b7-777292c2441d)


**Full**
Fresh environment
`pip install lightning deepspeed wandb neptune mlflow comet-ml tensorboard jsonargparse[signatures]`

![full-install](https://github.com/Lightning-AI/lightning/assets/5495193/8cbf379a-75df-4d64-b631-00dc4702ba85)


The raw files:

[import-minimal.log](https://github.com/Lightning-AI/lightning/files/12570155/import-minimal.log)
[import.log](https://github.com/Lightning-AI/lightning/files/12570156/import.log)


In summary, the lightning.app dependency imports is gone. We're good on the minimal-install benchmark since there the torch import is dominating. Some work can be done to reduce the import overhead in the loggers module. Also, TorchMetrics is importing torchvision in some metrics which could be avoided/delayed I guess. 

Overall, we are in a much better place now! 🎉
